kafka/producer-manager.go

```diff
@@ -0,0 +1,144 @@
+package kafka
+
+import (
+	"sync"
+
+	"time"
+
+	"net/http"
+
+	"github.com/Shopify/sarama"
+	"github.com/Sirupsen/logrus"
+	"github.com/pressly/chi"
+	"github.com/thrawn01/args"
+	"golang.org/x/net/context"
+)
+
+type contextKey int
+
+const (
+	producerManagerKey contextKey = 1
+)
+
+func SetProducerManager(ctx context.Context, manager *ProducerManager) context.Context {
+	return context.WithValue(ctx, producerManagerKey, manager)
+}
+
+func GetProducerManager(ctx context.Context) *ProducerManager {
+	obj, ok := ctx.Value(producerManagerKey).(*ProducerManager)
+	if !ok {
+		panic("No kafka.ProducerManager found in context")
+	}
+	return obj
+}
+
+type ProducerManager struct {
+	done      chan struct{}
+	parser    *args.ArgParser
+	current   chan Producer
```

Comment:
Is using ``current``, ``new`` and ``reconnect`` channels for reconnecting a known pattern? You used it for reconnecting to rethinkdb as well. Maybe encapsulate this common logic in a separate class? Isn't there a built-in reconnect logic in rethinkdb and kafka libs you used? Wouldn't it be easier to get a new session in reconnect, close the old one and remember the new one in a variable, called e.g. ``self.session``